### PR TITLE
DEV-203: Indexd deleted docs gap

### PR DESCRIPTION
**Pull request by mpsolano**
Includes the following changes from the ticket:

- Added not_deleted flag to get_all_versions() and get_latest_version() to filter out deleted documents.
- Added skip_deleted flag to bulk_get_latest_versions() to filter out deleted documents
- Made not_deleted/skip_deleted default to True for updated endpoints
- Added corresponding test cases.
- Modified openapi spec (swagger.yaml) to include newly added flags as query parameters.

As part of the update to the openapi spec, these additional changes were made:

- Updated base url to point to indexd consul service and switched out github references to point to NCI-GDC repo instead of uc-cdis repo.
- Added option to use HTTP schema in addition to HTTPS so it can be run with our current index consul service setup.
- Fixed validation errors in swagger.yaml.
  - Pre-existing validation errors were found in the yaml file.
  - When viewing swagger yaml in an swagger-editor it showed that several operationIds in the file were not unique.
  - Updated the ids to be unique which generated slightly different method names in the swagger_client that gets built.
  - Updated several test cases to use the new method names in swagger_client.
- Updated swagger.yaml to use openapi spec 3.0.
  - The only change this required was to test_openapi which validates the yaml is valid.
  - The validator (swagger_spec_validator) could not validate a 3.0 spec so swapped that out with openapi_spec_validator.
- Added an openapi-index.html to openapids folder to create an interactive documentation page.
  - This page can be opened easily in any IDE w/ a built-in web server for previewing.
  - Uses rapidoc to generate the page which is very lightweight with minimal dependencies.
  - Previously tried this with swagger-ui but that wasn’t ideal since it required that the swagger-ui project dist directory be placed in our indexd project.
  - Also briefly looked at redoc but that doesn’t have any interactive functionality which I think is a useful feature.

### CODE REVIEW

File: tests/test_driver_alchemy_crud.py [hunk 1]
```diff
@@ -398,6 +399,41 @@ def test_driver_get_latest_version_with_no_record(index_driver, database_conn):
...
+            index_metadata = json.dumps({'deleted': 'true'})
+
+        database_conn.execute(make_sql_statement("""
+            INSERT INTO index_record(did, baseid, rev, form, size, created_date, updated_date, index_metadata) 
+            VALUES (?,?,?,?,?,?,?,?)
+        """, (did, baseid, rev, form, size, created_date, updated_date, index_metadata)))
+
+    record = index_driver.get_latest_version(did, not_deleted=True)
+
+    assert record['baseid'] == baseid, 'record baseid does not match'
```
Comment (XukunXiangUChicago) [inline tests/test_driver_alchemy_crud.py:432]: nit-picking: it seems that we prefer `"`
```suggestion
    assert record["baseid"] == baseid, "record baseid does not match"
```

File: tests/test_driver_alchemy_crud.py [hunk 2]
```diff
@@ -464,6 +500,48 @@ def test_driver_get_all_version_with_no_record(index_driver, database_conn):
...
+    records = index_driver.get_all_versions(did, not_deleted=True)
+    assert len(records) == len(non_deleted_dids), 'the number of records does not match'
+
+    for record in records.values():
+        assert record['baseid'] == baseid, 'record baseid does not match'
+        assert record['did'] in non_deleted_dids, 'unexpected record did found'
+        assert record['did'] not in deleted_dids, 'deleted record did found'
+        non_deleted_dids.remove(record['did'])
+
+    assert not non_deleted_dids, 'one or more expected dids not found'
```
Comment (XukunXiangUChicago) [inline tests/test_driver_alchemy_crud.py:542]: nit-picking:
```suggestion
    assert all([doc["baseid"] == baseid for doc in records.values()]), "record baseid does not match"
    assert set(doc["did"] for doc in records.values()) == set(non_deleted_dids), "record did does not match"
```

File: tests/test_driver_alchemy_crud.py (hunk 1 above)
Comment (mpsolano) [reply tests/test_driver_alchemy_crud.py:432]: Sounds good, I'm updating both tests accordingly.

File: tests/test_driver_alchemy_crud.py (hunk 2 above)
Comment (mpsolano) [reply tests/test_driver_alchemy_crud.py:542]: This is much cleaner than my original implementation. Thank you for this! I'm making the update.

File: tests/test_openapi.py [hunk 3]
```diff
@@ -1,21 +1,20 @@
...
-        url = 'file:/' + f.name + '#'
-        spec = yaml.safe_load(f)
-        if not isinstance(spec, dict):
-            raise SwaggerValidationError('root node is not a mapping')
-        # ensure the spec is valid JSON
-        spec = json.loads(json.dumps(spec))
-        validator = swagger_spec_validator.util.get_validator(spec, url)
-        validator.validate_spec(spec, url)
+    filename = pkg_resources.resource_filename(openapis.__name__, 'swagger.yaml')
+    spec, url = read_from_filename(filename)
```
Comment (kulgan) [inline tests/test_openapi.py:13]: 👍🏾

File: tests/test_openapi.py [hunk 4]
```diff
@@ -1,21 +1,20 @@
 import json
 import pkg_resources
 
-import swagger_spec_validator.util
-import yaml
-from swagger_spec_validator.common import SwaggerValidationError
+from openapi_spec_validator import validate_spec
+from openapi_spec_validator.exceptions import OpenAPIValidationError
+from openapi_spec_validator.readers import read_from_filename
```
Comment (kulgan) [inline tests/test_openapi.py:6]: We generally try to stick to import just the modules for clarity while reading code. Something like this
```suggestion
import openapi_spec_validator
from openapi_spec_validator import exceptions as specs_exceptions
from openapi_spec_validator import readers as specs_readers
```

**Review by kulgan — COMMENTED** [review]
Implementation looks good to me, but there are a couple more endpoints that might need to be covered too. For example the `get_urls` and the `get_index` endpoints.

**Review by kulgan — APPROVED** [review]
lgtm

**Pull request merged**